x/auth/vesting/cmd/vestcalc/vestcalc.go

```diff
@@ -0,0 +1,532 @@
+			amount = 0
+		}
+		newEvents = append(newEvents, e)
+	}
+	if amount != 0 {
+		// special case if all events are before the cliff
+		cliffEvent := event{Time: cliff, Amount: amount}
+		newEvents = append(newEvents, cliffEvent)
+	}
+	// integrity check
+	oldTotal := int64(0)
+	for _, e := range events {
+		oldTotal = oldTotal + e.Amount
+	}
+	newTotal := int64(0)
+	for _, e := range newEvents {
+		newTotal = newTotal + e.Amount
+	}
+	if oldTotal != newTotal {
+		return nil, fmt.Errorf("applying vesting cliff changed total from %d to %d", oldTotal, newTotal)
+	}
+	return newEvents, nil
+}
+
+// eventsToPeriods converts the events to periods with the given start time
+// and denomination.
+func eventsToPeriods(startTime time.Time, events []event, denom string) []period {
+	periods := []period{}
+	lastTime := startTime
+	for _, e := range events {
+		dur := e.Time.Sub(lastTime)
+		p := period{
+			Coins:  encodeCoins(e.Amount, denom),
+			Length: int64(dur.Seconds()),
+		}
+		periods = append(periods, p)
+		lastTime = e.Time
+	}
+	return periods
+}
+
+// periodsToEvents converts periods to events with the given start time.
+func periodsToEvents(startTime time.Time, periods []period) []event {
+	events := []event{}
+	lastTime := startTime
+	for _, p := range periods {
+		amount, _ := parseCoins(p.Coins)
+		newTime := lastTime.Add(time.Duration(p.Length) * time.Second)
+		e := event{
+			Time:   newTime,
+			Amount: amount,
+		}
+		events = append(events, e)
+		lastTime = newTime
+	}
+	return events
+}
+
+const day = 24 * time.Hour
+
+// formatDuration returns a duration in a string like "3d4h3m0.5s".
+// It follows time.Duration.String() except that it includes 24-hour days.
+// NOTE: Does not reflect daylight savings changes.
+func formatDuration(d time.Duration) string {
+	s := ""
+	if d < 0 {
+		d = -d
+		s = "-"
+	}
+	if d < day {
+		// handle several special cases
+		return s + d.String()
+	}
+	// Now we know days are the most significant unit,
+	// so all other units should be present
+	days := int64(d / day)
+	r := d % day // remainder
+	s = s + fmt.Sprint(days) + "d"
+	hours := int64(r / time.Hour)
+	r = r % time.Hour
+	s = s + fmt.Sprint(hours) + "h"
+	minutes := int64(r / time.Minute)
+	r = r % time.Minute
+	s = s + fmt.Sprint(minutes) + "m"
+	seconds := int64(r / time.Second)
+	r = r % time.Second
+	s = s + fmt.Sprint(seconds) // no suffix yet
+	if r != 0 {
+		// Follow normal Duration formatting, but need to avoid
+		// the special handling of fractional seconds.
+		r = r + time.Second
+		frac := r.String()
+		// append skipping the leading "1"
+		return s + frac[1:] // adds the suffix
+	}
+	return s + "s" // now the suffix
+}
+
+// maxTime gives the maximum of a set of times, or the zero time if empty.
+func maxTime(cliffs []time.Time) time.Time {
+	tm := time.Time{}
+	for _, c := range cliffs {
+		if c.After(tm) {
+			tm = c
+		}
+	}
+	return tm
+}
+
+var (
+	validDenoms = map[string]bool{"ubld": true}
```

Comment:
Another TODO: "use cosmos-sdk denom validation, and allow any denom that passes that validation".

Reply:
Done.